### ISSUE TITLE
Strange indentation on constructor calls

### ISSUE THREAD

**Issue by TWiStErRob** [pinterest/ktlint#2422]
## Expected Behavior
No changes. See repro for details.

## Observed Behavior
```
MyClass.kt:4:1: Unexpected indentation (4) (should be 8) (standard:indent)
MyClass.kt:5:1: Unexpected indentation (4) (should be 8) (standard:indent)
MyClass.kt:6:1: Unexpected indentation (0) (should be 4) (standard:indent)
```

<details><summary>Full log</summary>

```
13:54:38.746 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered RuleSetProviderV3 with id 'standard' in ktlint JAR
13:54:39.013 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'baseline' in ktlint JAR
13:54:39.013 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'plain' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'checkstyle' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'json' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'format' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'html' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'plain-summary' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader - Discovered ReporterProviderV2 with id 'sarif' in ktlint JAR
13:54:39.014 [main] DEBUG com.pinterest.ktlint.cli.internal.ReporterAggregator - Initializing "plain" reporter with {plain=true, color=false, color_name=DARK_GRAY, format=false}
13:54:39.033 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.api.KtLintRuleEngine - Starting with linting file 'MyClass.kt'
13:54:39.408 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.EditorConfigLoader - Effective editorconfig properties for file 'P:\temp\asdf\MyClass.kt':
	
13:54:39.419 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.rulefilter.RunAfterRuleFilter - Rule with id 'standard:argument-list-wrapping' should run after the rule with id 'standard:class-signature'. However, the latter rule is not loaded and is allowed to be ignored. For best results, it is advised load the rule.
13:54:39.419 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.rulefilter.RunAfterRuleFilter - Rule with id 'standard:indent' should run after the rule with id 'standard:class-signature'. However, the latter rule is not loaded and is allowed to be ignored. For best results, it is advised load the rule.
13:54:39.420 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.rulefilter.RunAfterRuleFilter - Rule with id 'standard:argument-list-wrapping' should run after the rule with id 'standard:class-signature'. However, the latter rule is not loaded and is allowed to be ignored. For best results, it is advised load the rule.
13:54:39.420 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.rulefilter.RunAfterRuleFilter - Rule with id 'standard:indent' should run after the rule with id 'standard:class-signature'. However, the latter rule is not loaded and is allowed to be ignored. For best results, it is advised load the rule.
13:54:39.420 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.rulefilter.RunAfterRuleFilter - Rule with id 'standard:indent' should run after the rule with id 'standard:class-signature'. However, the latter rule is not loaded and is allowed to be ignored. For best results, it is advised load the rule.
13:54:39.444 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.RuleProviderSorter - Rules will be executed in order below:
           - internal:ktlint-suppression, 
           - standard:annotation-spacing, 
           - standard:blank-line-before-declaration, 
           - standard:chain-wrapping, 
           - standard:class-naming, 
           - standard:colon-spacing, 
           - standard:comma-spacing, 
           - standard:comment-spacing, 
           - standard:comment-wrapping, 
           - standard:context-receiver-wrapping, 
           - standard:curly-spacing, 
           - standard:discouraged-comment-location, 
           - standard:dot-spacing, 
           - standard:double-colon-spacing, 
           - standard:enum-entry-name-case, 
           - standard:enum-wrapping, 
           - standard:filename, 
           - standard:final-newline, 
           - standard:fun-keyword-spacing, 
           - standard:function-naming, 
           - standard:function-return-type-spacing, 
           - standard:function-start-of-body-spacing, 
           - standard:function-type-reference-spacing, 
           - standard:if-else-bracing, 
           - standard:import-ordering, 
           - standard:kdoc-wrapping, 
           - standard:keyword-spacing, 
           - standard:modifier-list-spacing, 
           - standard:modifier-order, 
           - standard:multiline-expression-wrapping, 
           - standard:multiline-if-else, 
           - standard:no-blank-line-before-rbrace, 
           - standard:no-blank-line-in-list, 
           - standard:no-blank-lines-in-chained-method-calls, 
           - standard:no-consecutive-blank-lines, 
           - standard:no-consecutive-comments, 
           - standard:no-empty-class-body, 
           - standard:no-empty-file, 
           - standard:no-empty-first-line-in-class-body, 
           - standard:no-empty-first-line-in-method-block, 
           - standard:no-line-break-after-else, 
           - standard:no-line-break-before-assignment, 
           - standard:no-multi-spaces, 
           - standard:no-trailing-spaces, 
           - standard:no-unit-return, 
           - standard:no-unused-imports, 
           - standard:no-wildcard-imports, 
           - standard:nullable-type-spacing, 
           - standard:op-spacing, 
           - standard:package-name, 
           - standard:parameter-list-spacing, 
           - standard:parameter-list-wrapping, 
           - standard:parameter-wrapping, 
           - standard:paren-spacing, 
           - standard:property-naming, 
           - standard:property-wrapping, 
           - standard:range-spacing, 
           - standard:spacing-around-angle-brackets, 
           - standard:spacing-between-declarations-with-annotations, 
           - standard:spacing-between-declarations-with-comments, 
           - standard:spacing-between-function-name-and-opening-parenthesis, 
           - standard:statement-wrapping, 
           - standard:string-template, 
           - standard:try-catch-finally-spacing, 
           - standard:type-argument-list-spacing, 
           - standard:type-parameter-list-spacing, 
           - standard:unary-op-spacing, 
           - standard:unnecessary-parentheses-before-trailing-lambda, 
           - standard:annotation, 
           - standard:if-else-wrapping, 
           - standard:no-single-line-block-comment, 
           - standard:wrapping, 
           - standard:argument-list-wrapping, 
           - standard:no-semi, 
           - standard:function-signature, 
           - standard:trailing-comma-on-call-site, 
           - standard:trailing-comma-on-declaration-site, 
           - standard:indent, 
           - standard:block-comment-initial-star-alignment, 
           - standard:string-template-indent, 
           - standard:max-line-length
13:54:39.513 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.api.KtLintRuleEngine - Finished with linting file 'MyClass.kt'
MyClass.kt:4:1: Unexpected indentation (4) (should be 8) (standard:indent)
MyClass.kt:5:1: Unexpected indentation (4) (should be 8) (standard:indent)
MyClass.kt:6:1: Unexpected indentation (0) (should be 4) (standard:indent)
13:54:39.516 [main] WARN com.pinterest.ktlint.cli.internal.KtlintCommandLine - Lint has found errors than can be autocorrected using 'ktlint --format'

Summary error count (descending) by rule:
  standard:indent: 3
13:54:39.518 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintCommandLine - Finished processing in 784ms / 1 file(s) scanned / 3 error(s) found
13:54:39.519 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintCommandLine - Exit ktlint with exit code: 1
```

</details>

## Steps to Reproduce
`ktlint "MyClass.kt" --log-level=debug --relative`

```kotlin
class MyClass(
    private val primaryProperty: ResourceFolderRepository,
) : SuperClass(
    superProperty = "super",
    superProperty2 = "super2",
) {
    private val privateProperty: File = primaryProperty.resourceDir

    fun doStuff() {
        val file = privateProperty.resolve("foo")
    }
}
```
ktlint wants this, which to me looks like the indentation is messed up (not aligned, not symmetric).
```kotlin
class MyClass(
    private val primaryProperty: ResourceFolderRepository,
) : SuperClass(
        superProperty = "super",
        superProperty2 = "super2",
    ) {
    private val privateProperty: File = primaryProperty.resourceDir

    fun doStuff() {
        val file = privateProperty.resolve("foo")
    }
}
```
also note that this code is valid (and looks good too):
```kotlin
class MyClass(private val primaryProperty: ResourceFolderRepository) : SuperClass(
    superProperty = "super",
    superProperty2 = "super2",
) {
    private val privateProperty: File = primaryProperty.resourceDir

    fun doStuff() {
        val file = privateProperty.resolve("foo")
    }
}
```

## Your Environment

* Version of ktlint used: 1.0.1
* Relevant parts of the `.editorconfig` settings: none
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): ktlint CLI
* Version of Gradle used (if applicable): N/A
* Operating System and version: Windows 10

**Comment by paul-dingemans** [comment]
Your `SuperClass` is a multiline construct, and as of that it has an extra indentation
```
SuperClass(
    superProperty = "super",
    superProperty2 = "super2",
)
```
`ktlint_official` code style favours to push down multiline constructs to a separate line. As that would look really weird for the super classes, it has been joined on the same line as the preceding `:`. Conceptually it is nested one level deeper though. As of that the parenthesis of the Superclass are symmetric and on the same level. But indeed, this different from what you may be used to compared to previous ktlint versions.

In case the code was written as:
```
class MyClass(
    private val primaryProperty: ResourceFolderRepository,
) : SuperClass(superProperty = "super", superProperty2 = "super2") {
    private val privateProperty: File = primaryProperty.resourceDir

    fun doStuff() {
        val file = privateProperty.resolve("foo")
    }
}
```
the `SuperClass(superProperty = "super", superProperty2 = "super2")` is a single line construct and formatting is unchanged.

Please note that you specified that you do not have an `.editorconfig`. In ktlint `1.0` the default code style has been changed from `intellij_idea` to `ktlint_official`. See https://pinterest.github.io/ktlint/latest/rules/code-styles/

**Comment by TWiStErRob** [comment]
Thanks for the details!

What is the actual difference between idea and official wrt this issue? Because the actual codes usually has more than 2 params and does not fit on one 140-long line. Is there a way to use official with this part of indent disabled?

**Comment by paul-dingemans** [comment]
Please see [example in docs](https://pinterest.github.io/ktlint/latest/rules/experimental/#class-signature). No, this can not be configured (see [faq](https://pinterest.github.io/ktlint/latest/faq/#can-a-new-toggle-be-added-to-optionally-enabledisable-format-code-in-a-particular-way)).

**Comment by TWiStErRob** [comment]
Ok, I guess this is WAI then and can be closed?

---

Either way, please note the following feedback.

The indentation aligns according to what you mentioned:
```
class MyClass(
    private val primaryProperty: ResourceFolderRepository,
) : |SuperClass(
    |    superProperty = "super",
    |    superProperty2 = "super2",
    |) {
    private val privateProperty: File = primaryProperty.resourceDir
```
this just happens to exactly be on a line (regardless of wrapping after `:` or not), because `) : ` is 4 characters long which matches the default indent level of 4.
If the indent is any other number of spaces, this becomes a bit more interesting, for example with 2 (often used alternative to 4):
```
class MyClass(
  private val primaryProperty: ResourceFolderRepository,
) : SuperClass(
    superProperty = "super",
    superProperty2 = "super2",
  ) {
  private val privateProperty: File = primaryProperty.resourceDir
```
in this case it does not align with "MyClass class declaration", neither with "SuperClass constructor call", it's indented one logical level inside, but it doesn't actually visually align with anything other than the class body, which makes it confusing.

You mentioned wrapping down super-invocation would look weird, I think this might be acceptable (it's different for sure, but IMO less strange than a "technically correct" indentation which visually only accidentally matches):
```
class MyClass(
    private val primaryProperty: ResourceFolderRepository,
) :
    SuperClass(
        superProperty = "super",
        superProperty2 = "super2",
    ),
    SomeInterface,
    OtherInterface,
{
    private val privateProperty: File = primaryProperty.resourceDir
    ...
}
```

**Comment by paul-dingemans** [comment]
> this just happens to exactly be on a line (regardless of wrapping after `:` or not), because `) : ` is 4 characters long which matches the default indent level of 4.
> If the indent is any other number of spaces, this becomes a bit more interesting, for example with 2 (often used alternative to 4):

This is a true observation. It holds true in more situations where the formatters (Intellij IDEA and ktlint) rely on the combination of the indent size of 4 characters versus the length of keyswords like `fun`, `val`, `var` (3 character plus space) or `if (` being exactly 4 characters to align things nicely.

When using other indent sizes, those 'benefits' are no longer present.

It is impossible to write an anti-bikeshedding formatter that produces exactly the format that any random developer wants. That would requires an enormous amount of configuration options which goes against the ant-bikeshedding idea. I aim to provide the three code styles which each produce a reasonable acceptable formatting based on what the community wants provided that the default settings are used, and no rules are disabled. When going down the route of tweaking the default configuration, things can result in unintended side effects.
